// prettier-plugin-java/dist/index.cjs
let web_tree_sitter = require("web-tree-sitter");
let prettier = require("prettier");
let prettier_doc = require("prettier/doc");
//#region src/options.ts
var options_default = {
	arrowParens: {
		type: "choice",
		category: "Java",
		default: "always",
		choices: [{
			value: "always",
			description: ""
		}, {
			value: "avoid",
			description: ""
		}],
		description: "Include parentheses around a sole arrow function parameter."
	},
	trailingComma: {
		type: "choice",
		category: "Java",
		default: "all",
		choices: [
			{
				value: "all",
				description: ""
			},
			{
				value: "es5",
				description: ""
			},
			{
				value: "none",
				description: ""
			}
		],
		description: "Print trailing commas wherever possible when multi-line."
	},
	experimentalOperatorPosition: {
		type: "choice",
		category: "Java",
		default: "end",
		choices: [{
			value: "start",
			description: ""
		}, {
			value: "end",
			description: ""
		}],
		description: "Where to print operators when binary expressions wrap lines."
	}
};
//#endregion
//#region src/node-types.ts
const multiFieldsByType = {
	array_creation_expression: { dimensions: true },
	cast_expression: { type: true },
	constant_declaration: { declarator: true },
	exports_module_directive: { modules: true },
	field_declaration: { declarator: true },
	for_statement: {
		init: true,
		update: true
	},
	local_variable_declaration: { declarator: true },
	opens_module_directive: { modules: true },
	provides_module_directive: { provider: true },
	requires_module_directive: { modifiers: true },
	spread_parameter: { annotations: true }
};
//#endregion
//#region src/printers/helpers.ts
const { group: group$7, hardline: hardline$6, ifBreak: ifBreak$2, indent: indent$7, join: join$7, line: line$7, softline: softline$5 } = prettier_doc.builders;
const { mapDoc } = prettier_doc.utils;
function hasType(path, type) {
	return path.node.type === type;
}
function hasChild(path, fieldName) {
	return path.node[fieldName] != null;
}
function definedKeys(obj, options) {
	return (options ?? Object.keys(obj)).filter((key) => obj[key] !== void 0);
}
function printModifiers(path, print, annotationMode) {
	const modifiersIndex = path.node.namedChildren.findIndex(({ type }) => type === "modifiers");
	if (modifiersIndex === -1) return [];
	const separator = annotationMode === "avoidBreak" ? line$7 : annotationMode === "noBreak" || path.node.namedChildren[modifiersIndex].children.some(({ type }) => type !== "annotation" && type !== "marker_annotation") ? " " : hardline$6;
	return [path.call((modifiers) => print(modifiers, { annotationMode }), "namedChildren", modifiersIndex), separator];
}
function printValue(path) {
	return path.node.value;
}
function lineStartWithComments(node) {
	return node.comments?.length ? Math.min(node.start.row, node.comments[0].start.row) : node.start.row;
}
function lineEndWithComments(node) {
	return node.comments?.length ? Math.max(node.end.row, node.comments.at(-1).end.row) : node.end.row;
}
function printDanglingComments(path) {
	if (!path.node.comments?.length) return [];
	const comments = [];
	path.each((commentPath) => {
		const comment = commentPath.node;
		if (comment.leading || comment.trailing) return;
		comment.printed = true;
		comments.push(printComment(comment));
	}, "comments");
	return join$7(hardline$6, comments);
}
function printComment(comment) {
	const lines = comment.value.split("\n").map((line) => line.trim());
	return lines.length > 1 && lines[0].startsWith("/*") && lines.slice(1).every((line) => line.startsWith("*")) && lines.at(-1).endsWith("*/") ? join$7(hardline$6, lines.map((line, index) => index === 0 ? line : ` ${line}`)) : comment.value;
}
function hasLeadingComments(node) {
	return node.comments?.some(({ leading }) => leading) ?? false;
}
function indentInParentheses(contents) {
	return !Array.isArray(contents) || contents.length ? [
		"(",
		indent$7([softline$5, contents]),
		softline$5,
		")"
	] : "()";
}
function printArrayInitializer(path, print, options) {
	if (!path.node.namedChildren.length) {
		const danglingComments = printDanglingComments(path);
		return danglingComments.length ? [
			"{",
			indent$7([hardline$6, ...danglingComments]),
			hardline$6,
			"}"
		] : "{}";
	}
	const list = join$7([",", line$7], path.map(print, "namedChildren"));
	if (list.length && options.trailingComma !== "none") list.push(ifBreak$2(","));
	return group$7([
		"{",
		indent$7([line$7, ...list]),
		line$7,
		"}"
	]);
}
function printBlock(path, contents) {
	if (contents.length) return group$7([
		"{",
		indent$7([hardline$6, ...join$7(hardline$6, contents)]),
		hardline$6,
		"}"
	]);
	const danglingComments = printDanglingComments(path);
	if (danglingComments.length) return [
		"{",
		indent$7([hardline$6, ...danglingComments]),
		hardline$6,
		"}"
	];
	const parent = path.parent;
	const grandparent = path.grandparent;
	return parent?.type === "catch_clause" && (grandparent?.type === "try_statement" || grandparent?.type === "try_with_resources_statement") && grandparent.namedChildren.filter(({ type }) => type === "catch_clause").length === 1 && !grandparent.namedChildren.some(({ type }) => type === "finally_clause") || parent && [
		"for_statement",
		"do_statement",
		"enhanced_for_statement",
		"while_statement"
	].includes(parent.type) || [
		"annotation_type_body",
		"class_body",
		"constructor_body",
		"enum_body",
		"interface_body",
		"module_body",
		"record_pattern_body"
	].includes(path.node.type) || parent && [
		"block",
		"lambda_expression",
		"method_declaration",
		"static_initializer",
		"synchronized_statement"
	].includes(parent.type) ? "{}" : [
		"{",
		hardline$6,
		"}"
	];
}
function printBlockStatements(path, print) {
	const parts = [];
	path.each((child) => {
		const { node, previous } = child;
		if (node.type === "switch_label") return;
		const blankLine = parts.length && previous && lineStartWithComments(node) > lineEndWithComments(previous) + 1;
		const declaration = print(child);
		parts.push(blankLine ? [hardline$6, declaration] : declaration);
	}, "namedChildren");
	return parts;
}
function printBodyDeclarations(path, print, padFirst = false) {
	const isInterfaceBody = path.node.type === "interface_body";
	const isFormalParameters = path.node.type === "formal_parameters";
	const separator = isFormalParameters ? softline$5 : hardline$6;
	let previousRequiresPadding = padFirst;
	return path.map((child) => {
		const { node, previous } = child;
		const modifiers = node.namedChildren.find(({ type }) => type === "modifiers")?.children ?? [];
		const firstAnnotationIndex = modifiers.findIndex(({ type }) => type === "annotation" || type === "marker_annotation");
		const lastNonAnnotationIndex = modifiers.findLastIndex(({ type }) => type !== "annotation" && type !== "marker_annotation");
		const currentRequiresPadding = firstAnnotationIndex !== -1 && (!isFormalParameters && lastNonAnnotationIndex === -1 || firstAnnotationIndex < lastNonAnnotationIndex) || !isFormalParameters && node.type !== "constant_declaration" && node.type !== "enum_constant" && node.type !== "field_declaration" && !(isInterfaceBody && node.type === "method_declaration" && !node.bodyNode);
		const blankLine = previousRequiresPadding || previous && (currentRequiresPadding || lineStartWithComments(node) > lineEndWithComments(previous) + 1);
		previousRequiresPadding = currentRequiresPadding;
		const declaration = print(child);
		return blankLine ? [separator, declaration] : declaration;
	}, "namedChildren");
}
function printVariableDeclaration(path, print) {
	const declaration = printModifiers(path, print);
	declaration.push(path.call(print, "typeNode"), " ");
	const declarators = path.map(print, "declaratorNodes");
	if (declarators.length > 1 && path.node.declaratorNodes.some(({ valueNode }) => valueNode)) declaration.push(group$7(indent$7(join$7([",", line$7], declarators)), { shouldBreak: path.parent?.type !== "for_statement" }));
	else declaration.push(join$7(", ", declarators));
	declaration.push(";");
	return declaration;
}
function printTextBlock(path, contents) {
	const parts = [
		"\"\"\"",
		hardline$6,
		contents,
		"\"\"\""
	];
	const parentType = path.parent?.type;
	const grandparentType = path.grandparent?.type;
	return parentType === "assignment_expression" || parentType === "variable_declarator" || path.node.fieldName === "object" && (grandparentType === "assignment_expression" || grandparentType === "variable_declarator") ? indent$7(parts) : parts;
}
function embedTextBlock(path) {
	if (path.node.namedChildren.some(({ type }) => type === "string_interpolation") || path.node.children[0].value === "\"") return null;
	const language = findEmbeddedLanguage(path);
	if (!language) return null;
	const text = unescapeTextBlockContents(textBlockContents(path.node));
	return async (textToDoc) => {
		return printTextBlock(path, [escapeDocForTextBlock(await textToDoc(text, { parser: language })), hardline$6]);
	};
}
function textBlockContents(node) {
	const lines = node.value.split("\n").slice(1);
	const baseIndent = findBaseIndent(lines);
	return lines.map((line) => line.slice(baseIndent)).join("\n").slice(0, -3);
}
function findBaseIndent(lines) {
	return Math.min(...lines.map((line) => line.search(/\S/)).filter((indent) => indent >= 0));
}
function findEmbeddedLanguage(path) {
	return path.ancestors.find(({ type, comments }) => type === "block" || comments?.some(({ leading }) => leading))?.comments?.filter(({ leading }) => leading).map(({ value }) => value.match(/^(?:\/\/|\/\*)\s*language\s*=\s*(\S+)/)?.[1]).findLast((language) => language)?.toLowerCase();
}
function escapeDocForTextBlock(doc) {
	return mapDoc(doc, (currentDoc) => typeof currentDoc === "string" ? currentDoc.replace(/\\|"""/g, (match) => match === "\\" ? "\\\\" : "\"\"\\\"") : currentDoc);
}
function unescapeTextBlockContents(text) {
	return text.replace(/\\(?:([stnr"'\\])|\n|\r\n?)/g, (_, escaped) => {
		switch (escaped) {
			case "s": return " ";
			case "t": return "	";
			case "n": return "\n";
			case "r": return "\r";
			default: return escaped ?? "";
		}
	});
}
//#endregion
//#region src/printers/arrays.ts
var arrays_default = { array_initializer: printArrayInitializer };
//#endregion
//#region src/printers/blocks-and-statements.ts
const { group: group$6, hardline: hardline$5, ifBreak: ifBreak$1, indent: indent$6, indentIfBreak: indentIfBreak$2, join: join$6, line: line$6, softline: softline$4 } = prettier_doc.builders;
var blocks_and_statements_default = {
	block(path, print) {
		return printBlock(path, printBlockStatements(path, print));
	},
	local_variable_declaration: printVariableDeclaration,
	labeled_statement(path, print) {
		return join$6(": ", path.map(print, "namedChildren"));
	},
	expression_statement(path, print) {
		const parentType = path.parent?.type;
		const expressionType = path.node.namedChildren[0].type;
		const expression = path.call(print, "namedChildren", 0);
		return expressionType === "switch_expression" && parentType !== "assignment_expression" && parentType !== "switch_rule" ? expression : [expression, ";"];
	},
	if_statement(path, print) {
		const statement = ["if ", path.call(print, "conditionNode")];
		if (path.node.consequenceNode.type === ";") statement.push(";");
		else statement.push(" ", path.call(print, "consequenceNode"));
		if (!hasChild(path, "alternativeNode")) return statement;
		const danglingComments = printDanglingComments(path);
		if (danglingComments.length) statement.push(hardline$5, ...danglingComments, hardline$5);
		else {
			const ifHasBlock = path.node.consequenceNode.type === "block";
			statement.push(ifHasBlock ? " " : hardline$5);
		}
		statement.push("else");
		if (path.node.alternativeNode.type === ";") statement.push(";");
		else statement.push(" ", path.call(print, "alternativeNode"));
		return statement;
	},
	assert_statement(path, print) {
		return [
			"assert ",
			...join$6(" : ", path.map(print, "namedChildren")),
			";"
		];
	},
	switch_expression(path, print) {
		return join$6(" ", [
			"switch",
			path.call(print, "conditionNode"),
			path.call(print, "bodyNode")
		]);
	},
	switch_block(path, print) {
		return printBlock(path, path.map(print, "namedChildren"));
	},
	switch_block_statement_group(path, print) {
		const parts = [];
		path.each((child) => {
			if (child.node.type === "switch_label") parts.push(print(child), ":");
		}, "namedChildren");
		const firstStatementIndex = path.node.namedChildren.findIndex(({ type }) => type !== "switch_label");
		if (firstStatementIndex === path.node.namedChildren.length - 1 && path.node.namedChildren[firstStatementIndex].type === "block") parts.push(" ", path.call(print, "namedChildren", firstStatementIndex));
		else if (firstStatementIndex !== -1) parts.push(indent$6([hardline$5, ...join$6(hardline$5, printBlockStatements(path, print))]));
		return parts;
	},
	switch_label(path, print) {
		if (!path.node.children.some(({ type }) => type === "case")) return "default";
		const values = [];
		path.each((child) => {
			if (child.node.type !== "guard") values.push(print(child));
		}, "namedChildren");
		const hasMultipleValues = values.length > 1;
		const label = hasMultipleValues ? ["case", indent$6([line$6, ...join$6([",", line$6], values)])] : ["case ", values[0]];
		const guardIndex = path.node.namedChildren.findIndex(({ type }) => type === "guard");
		return guardIndex !== -1 ? [group$6([...label, hasMultipleValues ? line$6 : " "]), path.call(print, "namedChildren", guardIndex)] : group$6(label);
	},
	switch_rule(path, print) {
		const bodyIndex = path.node.namedChildren.findIndex(({ type }) => type === "block" || type === "expression_statement" || type === "throw_statement");
		const body = path.call(print, "namedChildren", bodyIndex);
		const switchLabelIndex = path.node.namedChildren.findIndex(({ type }) => type === "switch_label");
		const parts = [path.call(print, "namedChildren", switchLabelIndex), " ->"];
		const bodyNode = path.node.namedChildren[bodyIndex];
		if (bodyNode.type !== "block" && hasLeadingComments(bodyNode)) parts.push(indent$6([hardline$5, body]));
		else parts.push(" ", body);
		return parts;
	},
	while_statement(path, print) {
		const parts = ["while ", path.call(print, "conditionNode")];
		const body = path.call(print, "bodyNode");
		const bodyType = path.node.bodyNode.type;
		if (bodyType === "block") {
			parts.push(" ", body);
			return parts;
		} else if (bodyType === ";") {
			parts.push(";");
			return parts;
		} else {
			parts.push(line$6, body);
			return group$6(indent$6(parts));
		}
	},
	do_statement(path, print) {
		return [
			"do",
			path.node.bodyNode.type === ";" ? ";" : [" ", path.call(print, "bodyNode")],
			" while ",
			path.call(print, "conditionNode"),
			";"
		];
	},
	for_statement(path, print) {
		const danglingComments = printDanglingComments(path);
		if (danglingComments.length) danglingComments.push(hardline$5);
		const hasInit = path.node.initNodes.length > 0;
		const hasCondition = hasChild(path, "conditionNode");
		const hasUpdate = path.node.updateNodes.length > 0;
		const expressions = [
			!hasInit ? ";" : path.node.initNodes[0].type === "local_variable_declaration" ? path.call(print, "initNodes", 0) : [printExpressionList(path.map(print, "initNodes")), ";"],
			hasCondition ? [path.call(print, "conditionNode"), ";"] : ";",
			hasUpdate ? printExpressionList(path.map(print, "updateNodes")) : ""
		];
		const hasEmptyStatement = path.node.bodyNode.type === ";";
		return [
			...danglingComments,
			"for ",
			hasInit || hasCondition || hasUpdate ? group$6(indentInParentheses(join$6(line$6, expressions))) : "(;;)",
			hasEmptyStatement ? ";" : [" ", path.call(print, "bodyNode")]
		];
	},
	update_expression(path, print) {
		return path.map(print, "children");
	},
	enhanced_for_statement(path, print) {
		const forStatement = printDanglingComments(path);
		forStatement.push("for (", ...printModifiers(path, print), path.call(print, "typeNode"), " ", path.call(print, "nameNode"));
		if (hasChild(path, "dimensionsNode")) forStatement.push(path.call(print, "dimensionsNode"));
		forStatement.push(" : ", path.call(print, "valueNode"), ")");
		const bodyType = path.node.bodyNode.type;
		if (bodyType === ";") forStatement.push(";");
		else {
			const body = path.call(print, "bodyNode");
			forStatement.push(bodyType === "block" ? [" ", body] : indent$6([line$6, body]));
		}
		return group$6(forStatement);
	},
	break_statement(path, print) {
		const parts = ["break"];
		const identifierIndex = path.node.namedChildren.findIndex(({ type }) => type === "identifier");
		if (identifierIndex !== -1) parts.push(" ", path.call(print, "namedChildren", identifierIndex));
		parts.push(";");
		return parts;
	},
	continue_statement(path, print) {
		const identifierIndex = path.node.namedChildren.findIndex(({ type }) => type === "identifier");
		return identifierIndex !== -1 ? [
			"continue ",
			path.call(print, "namedChildren", identifierIndex),
			";"
		] : "continue;";
	},
	return_statement(path, print) {
		const statement = ["return"];
		if (path.node.namedChildren.length) {
			statement.push(" ");
			const expression = path.call(print, "namedChildren", 0);
			if (path.node.namedChildren[0].type === "binary_expression") statement.push(group$6([
				ifBreak$1("("),
				indent$6([softline$4, expression]),
				softline$4,
				ifBreak$1(")")
			]));
			else statement.push(expression);
		}
		statement.push(";");
		return statement;
	},
	throw_statement(path, print) {
		return [
			"throw ",
			path.call(print, "namedChildren", 0),
			";"
		];
	},
	synchronized_statement(path, print) {
		const parenthesizedExpressionIndex = path.node.namedChildren.findIndex(({ type }) => type === "parenthesized_expression");
		return [
			"synchronized ",
			path.call(print, "namedChildren", parenthesizedExpressionIndex),
			" ",
			path.call(print, "bodyNode")
		];
	},
	try_statement(path, print) {
		const parts = ["try", path.call(print, "bodyNode")];
		path.each((child) => {
			if (child.node.type === "catch_clause" || child.node.type === "finally_clause") parts.push(print(child));
		}, "namedChildren");
		return join$6(" ", parts);
	},
	catch_clause(path, print) {
		const catchFormalParameterIndex = path.node.namedChildren.findIndex(({ type }) => type === "catch_formal_parameter");
		return [
			"catch ",
			group$6(indentInParentheses(path.call(print, "namedChildren", catchFormalParameterIndex))),
			" ",
			path.call(print, "bodyNode")
		];
	},
	catch_formal_parameter(path, print) {
		const parts = printModifiers(path, print, "noBreak");
		const catchTypeIndex = path.node.namedChildren.findIndex(({ type }) => type === "catch_type");
		parts.push(path.call(print, "namedChildren", catchTypeIndex), " ", path.call(print, "nameNode"));
		if (hasChild(path, "dimensionsNode")) parts.push(path.call(print, "dimensionsNode"));
		return parts;
	},
	catch_type(path, print) {
		return join$6([line$6, "| "], path.map(print, "namedChildren"));
	},
	finally_clause(path, print) {
		return ["finally ", path.call(print, "namedChildren", 0)];
	},
	try_with_resources_statement(path, print) {
		const parts = [
			"try",
			path.call(print, "resourcesNode"),
			path.call(print, "bodyNode")
		];
		path.each((child) => {
			if (child.node.type === "catch_clause" || child.node.type === "finally_clause") parts.push(print(child));
		}, "namedChildren");
		return join$6(" ", parts);
	},
	resource_specification(path, print) {
		const resources = [];
		let hasTrailingSemicolon = false;
		path.each((child) => {
			if (child.node.type === "resource") {
				resources.push(print(child));
				hasTrailingSemicolon = false;
			} else if (child.node.type === ";") hasTrailingSemicolon = true;
		}, "children");
		const parts = join$6([";", line$6], resources);
		if (hasTrailingSemicolon) parts.push(ifBreak$1(";"));
		return group$6(indentInParentheses(parts));
	},
	resource(path, print) {
		if (hasChild(path, "typeNode") && hasChild(path, "nameNode") && hasChild(path, "valueNode")) {
			const parts = printModifiers(path, print);
			parts.push(path.call(print, "typeNode"), " ", path.call(print, "nameNode"));
			if (hasChild(path, "dimensionsNode")) parts.push(path.call(print, "dimensionsNode"));
			parts.push(" =");
			const value = path.call(print, "valueNode");
			if (path.node.valueNode.type === "binary_expression" || hasLeadingComments(path.node.valueNode)) parts.push(group$6(indent$6([line$6, value])));
			else {
				const groupId = Symbol("assignment");
				parts.push(group$6(indent$6(line$6), { id: groupId }), indentIfBreak$2(value, { groupId }));
			}
			return parts;
		}
		const resourceIndex = path.node.namedChildren.findIndex(({ type }) => type === "identifier" || type === "field_access");
		return path.call(print, "namedChildren", resourceIndex);
	},
	yield_statement(path, print) {
		return [
			"yield ",
			path.call(print, "namedChildren", 0),
			";"
		];
	}
};
function printExpressionList(expressions) {
	return group$6(expressions.map((expression, index) => index === 0 ? expression : [",", indent$6([line$6, expression])]));
}
//#endregion
//#region src/printers/classes.ts
const { group: group$5, hardline: hardline$4, indent: indent$5, indentIfBreak: indentIfBreak$1, join: join$5, line: line$5, softline: softline$3 } = prettier_doc.builders;
var classes_default = {
	class_declaration(path, print) {
		const parts = ["class ", path.call(print, "nameNode")];
		const definedClauses = definedKeys(path.node, [
			"superclassNode",
			"interfacesNode",
			"permitsNode"
		]);
		const hasMultipleClauses = definedClauses.length > 1;
		const hasTypeParameters = hasChild(path, "type_parametersNode");
		if (hasTypeParameters) {
			const typeParameters = path.call(print, "type_parametersNode");
			parts.push(hasMultipleClauses ? group$5(indent$5(typeParameters)) : typeParameters);
		}
		if (definedClauses.length) {
			const separator = hasTypeParameters && !hasMultipleClauses ? " " : line$5;
			const clauses = definedClauses.flatMap((clause) => hasChild(path, clause) ? [separator, path.call(print, clause)] : []);
			const hasBody = path.node.bodyNode.namedChildren.length > 0;
			const clauseGroup = [hasTypeParameters && !hasMultipleClauses ? clauses : indent$5(clauses), hasBody ? separator : " "];
			parts.push(hasMultipleClauses ? clauseGroup : group$5(clauseGroup));
		} else parts.push(" ");
		return [
			...printModifiers(path, print, "declarationOnly"),
			group$5(parts),
			path.call(print, "bodyNode")
		];
	},
	type_parameters(path, print) {
		return [
			"<",
			indent$5([softline$3, join$5([",", line$5], path.map(print, "namedChildren"))]),
			softline$3,
			">"
		];
	},
	superclass(path, print) {
		return ["extends ", path.call(print, "namedChildren", 0)];
	},
	super_interfaces(path, print) {
		return group$5(["implements", indent$5([line$5, path.call(print, "namedChildren", 0)])]);
	},
	permits(path, print) {
		return group$5(["permits", indent$5([line$5, path.call(print, "namedChildren", 0)])]);
	},
	type_list(path, print) {
		return join$5([",", line$5], path.map(print, "namedChildren"));
	},
	class_body(path, print) {
		return printBlock(path, printBodyDeclarations(path, print, path.parent?.type === "class_declaration"));
	},
	field_declaration: printVariableDeclaration,
	variable_declarator(path, print) {
		const declarator = [path.call(print, "nameNode")];
		if (hasChild(path, "dimensionsNode")) declarator.push(path.call(print, "dimensionsNode"));
		if (!hasChild(path, "valueNode")) return declarator;
		declarator.push(" =");
		const value = path.call(print, "valueNode");
		if (path.node.valueNode.type === "binary_expression" || path.node.valueNode.type === "ternary_expression" && path.node.valueNode.conditionNode.type === "binary_expression" || hasLeadingComments(path.node.valueNode)) declarator.push(group$5(indent$5([line$5, value])));
		else {
			const groupId = Symbol("assignment");
			declarator.push(group$5(indent$5(line$5), { id: groupId }), indentIfBreak$1(value, { groupId }));
		}
		return group$5(declarator);
	},
	method_declaration(path, print) {
		const modifiers = printModifiers(path, print);
		const declaration = [];
		if (hasChild(path, "type_parametersNode")) declaration.push(group$5(path.call(print, "type_parametersNode")), " ");
		path.each((child) => {
			if (child.node.type === "annotation" || child.node.type === "marker_annotation") declaration.push(print(child), " ");
		}, "children");
		declaration.push(path.call(print, "typeNode"));
		if (hasChild(path, "dimensionsNode")) declaration.push(path.call(print, "dimensionsNode"));
		declaration.push(" ", path.call(print, "nameNode"), group$5(path.call(print, "parametersNode")));
		const throwsIndex = path.node.namedChildren.findIndex(({ type }) => type === "throws");
		if (throwsIndex !== -1) declaration.push(group$5(indent$5([line$5, path.call(print, "namedChildren", throwsIndex)])));
		return hasChild(path, "bodyNode") ? [
			modifiers,
			group$5(declaration),
			" ",
			path.call(print, "bodyNode")
		] : [
			modifiers,
			group$5(declaration),
			";"
		];
	},
	receiver_parameter(path, print) {
		return path.map((child) => child.isLast ? print(child) : [print(child), child.node.type === "identifier" ? "." : " "], "namedChildren");
	},
	formal_parameters(path, print) {
		return indentInParentheses(join$5([",", line$5], path.parent?.type === "record_declaration" ? printBodyDeclarations(path, print) : path.map(print, "namedChildren")));
	},
	formal_parameter(path, print) {
		const parameter = printModifiers(path, print, path.grandparent?.type === "record_declaration" ? "avoidBreak" : "noBreak");
		parameter.push(path.call(print, "typeNode"));
		if (hasChild(path, "dimensionsNode")) parameter.push(path.call(print, "dimensionsNode"));
		parameter.push(" ", path.call(print, "nameNode"));
		return group$5(parameter);
	},
	spread_parameter(path, print) {
		const parts = printModifiers(path, print, "noBreak");
		parts.push(path.call(print, "typeNode"));
		if (hasChild(path, "annotationsNodes")) parts.push(...path.map((annotation) => [" ", print(annotation)], "annotationsNodes"));
		parts.push("... ", path.call(print, "namedChildren", path.node.namedChildren.length - 1));
		return parts;
	},
	throws(path, print) {
		return ["throws ", ...join$5(", ", path.map(print, "namedChildren"))];
	},
	static_initializer(path, print) {
		return ["static ", path.call(print, "namedChildren", 0)];
	},
	constructor_declaration(path, print) {
		const modifiers = printModifiers(path, print, "declarationOnly");
		const declaration = [];
		if (hasChild(path, "type_parametersNode")) declaration.push(group$5(path.call(print, "type_parametersNode")), " ");
		declaration.push(path.call(print, "nameNode"), group$5(path.call(print, "parametersNode")));
		const throwsIndex = path.node.namedChildren.findIndex(({ type }) => type === "throws");
		if (throwsIndex !== -1) declaration.push(group$5(indent$5([line$5, path.call(print, "namedChildren", throwsIndex)])));
		return [
			modifiers,
			group$5(declaration),
			" ",
			path.call(print, "bodyNode")
		];
	},
	constructor_body(path, print) {
		return printBlock(path, printBlockStatements(path, print));
	},
	explicit_constructor_invocation(path, print) {
		const invocation = [];
		if (hasChild(path, "objectNode")) invocation.push(path.call(print, "objectNode"), ".");
		if (hasChild(path, "type_argumentsNode")) invocation.push(path.call(print, "type_argumentsNode"));
		invocation.push(path.call(print, "constructorNode"), path.call(print, "argumentsNode"), ";");
		return invocation;
	},
	visibility: printValue,
	modifier: printValue,
	modifiers(path, print, _, args) {
		const parts = [];
		const modifiers = [];
		const typeAnnotations = [];
		path.each((child) => {
			if (child.node.type === "annotation" || child.node.type === "marker_annotation") (modifiers.length ? typeAnnotations : parts).push(print(child));
			else {
				modifiers.push(child.node.value);
				parts.push(...typeAnnotations);
				typeAnnotations.length = 0;
			}
		}, "namedChildren");
		const annotationMode = args && typeof args === "object" && "annotationMode" in args ? args.annotationMode : null;
		if (annotationMode === "declarationOnly") {
			parts.push(...typeAnnotations);
			typeAnnotations.length = 0;
		}
		modifiers.sort((a, b) => (indexByModifier.get(a) ?? -1) - (indexByModifier.get(b) ?? -1));
		if (modifiers.length || typeAnnotations.length) parts.push(join$5(" ", [...modifiers, ...typeAnnotations]));
		return join$5(annotationMode === "avoidBreak" ? line$5 : annotationMode === "noBreak" ? " " : hardline$4, parts);
	},
	enum_declaration(path, print) {
		const parts = printModifiers(path, print);
		parts.push("enum ", path.call(print, "nameNode"));
		if (hasChild(path, "interfacesNode")) {
			const hasBody = path.node.bodyNode.namedChildren.length > 0;
			parts.push(indent$5([line$5, path.call(print, "interfacesNode")]), hasBody ? line$5 : " ");
		} else parts.push(" ");
		return [group$5(parts), path.call(print, "bodyNode")];
	},
	enum_body(path, print, options) {
		const parts = printBodyDeclarations(path, print);
		const enumBodyDeclarationsIndex = path.node.namedChildren.findIndex(({ type }) => type === "enum_body_declarations");
		const declarations = [];
		if (enumBodyDeclarationsIndex !== -1) {
			const hasDeclarations = path.node.namedChildren[enumBodyDeclarationsIndex].namedChildren.length > 0;
			const enumBodyDeclarations = parts.pop();
			if (hasDeclarations) {
				if (!parts.length) declarations.push(hardline$4);
				declarations.push(enumBodyDeclarations);
			}
		}
		const contents = [];
		if (parts.length) {
			contents.push(join$5([",", hardline$4], parts));
			if (!declarations.length && options.trailingComma !== "none") contents.push(",");
		}
		if (declarations.length) contents.push(";", hardline$4, ...declarations);
		return printBlock(path, contents.length ? [contents] : []);
	},
	enum_constant(path, print) {
		const initializer = printModifiers(path, print);
		initializer.push(path.call(print, "nameNode"));
		if (hasChild(path, "argumentsNode")) initializer.push(path.call(print, "argumentsNode"));
		if (hasChild(path, "bodyNode")) initializer.push(" ", path.call(print, "bodyNode"));
		return initializer;
	},
	enum_body_declarations(path, print) {
		return join$5(hardline$4, printBodyDeclarations(path, print));
	},
	record_declaration(path, print) {
		const modifiers = printModifiers(path, print, "declarationOnly");
		const parts = ["record ", path.call(print, "nameNode")];
		if (hasChild(path, "type_parametersNode")) parts.push(group$5(path.call(print, "type_parametersNode")));
		parts.push(path.call(print, "parametersNode"));
		if (hasChild(path, "interfacesNode")) {
			const hasParameters = path.node.parametersNode.namedChildren.length > 0;
			const hasBody = path.node.bodyNode.namedChildren.length > 0;
			const interfaces = [hasParameters ? " " : line$5, path.call(print, "interfacesNode")];
			parts.push(group$5([hasParameters ? interfaces : indent$5(interfaces), hasBody ? line$5 : " "]));
		} else parts.push(" ");
		return [
			...modifiers,
			group$5(parts),
			path.call(print, "bodyNode")
		];
	},
	compact_constructor_declaration(path, print) {
		const parts = printModifiers(path, print, "declarationOnly");
		parts.push(path.call(print, "nameNode"), " ", path.call(print, "bodyNode"));
		return parts;
	}
};
const indexByModifier = [
	"public",
	"protected",
	"private",
	"abstract",
	"default",
	"static",
	"final",
	"transient",
	"volatile",
	"synchronized",
	"native",
	"sealed",
	"non-sealed",
	"strictfp"
].reduce((map, name, index) => map.set(name, index), /* @__PURE__ */ new Map());
//#endregion
//#region src/printers/expressions.ts
const { align, breakParent: breakParent$1, conditionalGroup, group: group$4, hardline: hardline$3, ifBreak, indent: indent$4, indentIfBreak, join: join$4, line: line$4, softline: softline$2 } = prettier_doc.builders;
const { getNextNonSpaceNonCommentCharacterIndex, hasNewline: hasNewline$1, isNextLineEmpty } = prettier.util;
const { removeLines, willBreak } = prettier_doc.utils;
var expressions_default = {
	lambda_expression(path, print, options, args) {
		const signatureDocs = [];
		let bodyDoc;
		const bodyComments = [];
		const shouldPrintAsChain = !(args && typeof args === "object" && "expandLastArg" in args && args.expandLastArg) && path.node.bodyNode.type === "lambda_expression";
		let functionBody;
		(function rec() {
			const { node } = path;
			const signatureDoc = printLambdaExpressionSignature(path, options, print, args);
			if (signatureDocs.length === 0) signatureDocs.push(signatureDoc);
			else {
				const { leading, trailing } = printCommentsSeparately(path);
				signatureDocs.push([leading, signatureDoc]);
				bodyComments.unshift(trailing);
			}
			if (!shouldPrintAsChain || node.bodyNode.type !== "lambda_expression") {
				bodyDoc = path.call((child) => print(child, args), "bodyNode");
				functionBody = node.bodyNode;
			} else path.call(rec, "bodyNode");
		})();
		const shouldPutBodyOnSameLine = !functionBody.comments?.some(({ leading }) => leading && hasNewline$1(options.originalText, functionBody.end.index)) && mayBreakAfterShortPrefix(functionBody);
		const isCallee = path.node.fieldName === "object" && path.parent?.type === "method_invocation";
		const chainGroupId = Symbol("arrow-chain");
		const signaturesDoc = printArrowFunctionSignatures(path, { signatureDocs });
		let shouldBreakSignatures = false;
		let shouldIndentSignatures = false;
		let shouldPrintSoftlineInIndent = false;
		if (shouldPrintAsChain && isCallee) {
			shouldIndentSignatures = true;
			shouldPrintSoftlineInIndent = !path.node.comments?.some(({ type, leading }) => leading && type === "line_comment");
			shouldBreakSignatures = isCallee && !shouldPutBodyOnSameLine;
		}
		const trailingSpace = args && typeof args === "object" && "expandLastArg" in args && args.expandLastArg && !path.node.comments ? softline$2 : "";
		bodyDoc = shouldPutBodyOnSameLine ? [
			" ",
			bodyDoc,
			bodyComments
		] : [indent$4([
			line$4,
			bodyDoc,
			bodyComments
		]), trailingSpace];
		return group$4([
			group$4(shouldIndentSignatures ? indent$4([shouldPrintSoftlineInIndent ? softline$2 : "", signaturesDoc]) : signaturesDoc, {
				shouldBreak: shouldBreakSignatures,
				id: chainGroupId
			}),
			" ->",
			shouldPrintAsChain ? indentIfBreak(bodyDoc, { groupId: chainGroupId }) : group$4(bodyDoc),
			shouldPrintAsChain && isCallee ? ifBreak(softline$2, "", { groupId: chainGroupId }) : ""
		]);
	},
	inferred_parameters(path, print, options) {
		const identifiers = [];
		path.each((child) => {
			if (child.node.type === "identifier") identifiers.push(print(child));
		}, "namedChildren");
		if (!identifiers.length) return "()";
		const parameters = join$4([",", line$4], identifiers);
		if (identifiers.length > 1) return group$4(indentInParentheses(parameters));
		return options.arrowParens === "avoid" ? parameters : [
			"(",
			...parameters,
			")"
		];
	},
	ternary_expression(path, print, options) {
		const condition = path.call(print, "conditionNode");
		const consequence = path.call(print, "consequenceNode");
		const alternative = path.call(print, "alternativeNode");
		const parentType = path.parent?.type;
		const isNestedTernary = parentType === "ternary_expression";
		const suffix = [
			line$4,
			["? ", options.useTabs ? indent$4(consequence) : align(2, consequence)],
			line$4,
			[": ", options.useTabs ? indent$4(alternative) : align(2, alternative)]
		];
		const prefix = group$4(condition);
		const alignedSuffix = !isNestedTernary || options.useTabs ? suffix : align(Math.max(0, options.tabWidth - 2), suffix);
		if (isNestedTernary) return [prefix, alignedSuffix];
		const parts = [prefix, indent$4(alignedSuffix)];
		return parentType === "parenthesized_expression" ? parts : group$4(parts);
	},
	assignment_expression(path, print) {
		const { operatorNode, rightNode } = path.node;
		const parts = [
			path.call(print, "leftNode"),
			" ",
			operatorNode.type
		];
		const right = path.call(print, "rightNode");
		if (rightNode.type === "binary_expression" || rightNode.type === "ternary_expression" && rightNode.conditionNode.type === "binary_expression" || hasLeadingComments(rightNode)) parts.push(group$4(indent$4([line$4, right])));
		else {
			const groupId = Symbol("assignment");
			parts.push(group$4(indent$4(line$4), { id: groupId }), indentIfBreak(right, { groupId }));
		}
		return parts;
	},
	binary_expression(path, print, options) {
		const { node } = path;
		const parent = path.parent;
		const grandparent = path.grandparent;
		const isInsideParentheses = parent?.fieldName === "condition" && (grandparent?.type === "if_statement" || grandparent?.type === "while_statement" || grandparent?.type === "switch_expression" || grandparent?.type === "do_statement") || parent?.fieldName !== "body" && grandparent?.type === "synchronized_statement";
		const parts = printBinaryExpressions(path, print, options, isInsideParentheses);
		if (isInsideParentheses) return parts;
		if (parent?.fieldName === "object" && (grandparent?.type === "method_invocation" || grandparent?.type === "explicit_constructor_invocation" || grandparent?.type === "field_access") || grandparent?.type === "method_reference") return parts;
		if (parent?.type === "return_statement" || parent?.type === "throw_statement" || parent?.type === "parenthesized_expression" || parent?.type === "assignment_expression" || parent?.type === "variable_declarator" || parent?.type === "guard" || node.fieldName === "body" && parent?.type === "lambda_expression" || node.fieldName !== "body" && parent?.type === "for_statement" || parent?.type === "ternary_expression" && grandparent?.type !== "return_statement" && grandparent?.type !== "throw_statement" && grandparent?.type !== "parenthesized_expression" && grandparent?.type !== "argument_list" || node.fieldName === "operand" && parent?.type === "unary_expression") return group$4(parts);
		if (parts.length === 0) return "";
		const firstGroupIndex = parts.findIndex((part) => typeof part !== "string" && !Array.isArray(part) && part.type === "group");
		const headParts = parts.slice(0, firstGroupIndex === -1 ? 1 : firstGroupIndex + 1);
		const rest = parts.slice(headParts.length);
		return group$4([...headParts, indent$4(rest)]);
	},
	instanceof_expression(path, print, options) {
		return group$4(indent$4(path.map((child) => {
			const doc = print(child, { annotationMode: "noBreak" });
			if (!child.previous) return doc;
			return [(options.experimentalOperatorPosition === "start" ? child.node.type === "instanceof" : child.previous.type === "instanceof") ? line$4 : " ", doc];
		}, "children")));
	},
	unary_expression(path, print) {
		return path.map(print, "children");
	},
	field_access: printMemberChain,
	generic_type(path, print) {
		const typeIdentifierIndex = path.node.namedChildren.findIndex(({ type }) => type === "scoped_type_identifier" || type === "type_identifier");
		const typeArgumentsIndex = path.node.namedChildren.findIndex(({ type }) => type === "type_arguments");
		return [path.call(print, "namedChildren", typeIdentifierIndex), path.call(print, "namedChildren", typeArgumentsIndex)];
	},
	parenthesized_expression(path, print) {
		const expression = path.call(print, "namedChildren", 0);
		const parentType = path.parent?.type;
		const grandparentType = path.grandparent?.type;
		const expressionType = path.node.namedChildren[0].type;
		const hasLambda = expressionType === "lambda_expression";
		const hasTernary = expressionType === "ternary_expression";
		const hasSuffix = parentType && (parentType === "array_access" || parentType === "explicit_constructor_invocation" || parentType === "field_access" || parentType === "method_invocation" || parentType === "method_reference" || parentType === "object_creation_expression");
		const isAssignment = parentType && (parentType === "assignment_expression" || parentType === "variable_declarator") || hasSuffix && (grandparentType === "assignment_expression" || grandparentType === "variable_declarator");
		if (!hasLambda && hasSuffix && (!hasTernary || isAssignment)) return group$4(indentInParentheses(hasTernary ? group$4(expression) : expression));
		else if (parentType && (parentType === "guard" || parentType === "return_statement" || parentType === "unary_expression" && grandparentType === "return_statement" || path.node.fieldName === "condition" && (parentType === "do_statement" || parentType === "if_statement" || parentType === "switch_expression" || parentType === "while_statement") || path.node.fieldName !== "body" && parentType === "synchronized_statement")) return group$4(indentInParentheses(group$4(expression)));
		else if (hasTernary && hasSuffix && !isAssignment) return group$4([
			"(",
			expression,
			softline$2,
			")"
		]);
		else return group$4([
			"(",
			hasLambda || hasTernary ? expression : indent$4(expression),
			")"
		]);
	},
	cast_expression(path, print) {
		const types = path.map(print, "typeNodes");
		const value = path.call(print, "valueNode");
		return types.length > 1 ? [
			group$4(indentInParentheses(join$4([line$4, "& "], types))),
			" ",
			value
		] : [
			"(",
			...types,
			") ",
			value
		];
	},
	object_creation_expression(path, print) {
		const expression = [];
		path.each((child) => {
			if (child.node.type === "class_body") expression.push(" ");
			expression.push(print(child));
			if (child.node.type === "annotation" || child.node.type === "marker_annotation" || child.node.type === "new") expression.push(" ");
		}, "children");
		return expression;
	},
	method_invocation: printMemberChain,
	argument_list(path, print, options) {
		const args = path.node.namedChildren;
		if (args.length === 0) {
			const shouldBreak = path.node.comments?.some(({ type, leading, trailing }) => !leading && !trailing && type === "line_comment");
			return group$4(indentInParentheses(printDanglingComments(path)), { shouldBreak });
		}
		const lastArgIndex = args.length - 1;
		let anyArgEmptyLine = false;
		const printedArguments = [];
		path.each((arg, index) => {
			let argDoc = print(arg);
			if (index === lastArgIndex) {} else if (isNextLineEmpty(options.originalText, arg.node.end.index)) {
				anyArgEmptyLine = true;
				argDoc = [
					argDoc,
					",",
					hardline$3,
					hardline$3
				];
			} else argDoc = [
				argDoc,
				",",
				line$4
			];
			printedArguments.push(argDoc);
		}, "namedChildren");
		function allArgsBrokenOut() {
			return group$4([
				"(",
				indent$4([line$4, ...printedArguments]),
				line$4,
				")"
			], { shouldBreak: true });
		}
		if (anyArgEmptyLine) return allArgsBrokenOut();
		if (shouldExpandFirstArg(args)) {
			const tailArgs = printedArguments.slice(1);
			if (tailArgs.some(willBreak)) return allArgsBrokenOut();
			let firstArg;
			try {
				firstArg = path.call((arg) => print(arg, { expandFirstArg: true }), "namedChildren", 0);
			} catch (caught) {
				if (caught instanceof ArgExpansionBailout) return allArgsBrokenOut();
				throw caught;
			}
			if (willBreak(firstArg)) return [breakParent$1, conditionalGroup([[
				"(",
				group$4(firstArg, { shouldBreak: true }),
				", ",
				...tailArgs,
				")"
			], allArgsBrokenOut()])];
			return conditionalGroup([
				[
					"(",
					firstArg,
					", ",
					...tailArgs,
					")"
				],
				[
					"(",
					group$4(firstArg, { shouldBreak: true }),
					", ",
					...tailArgs,
					")"
				],
				allArgsBrokenOut()
			]);
		}
		if (shouldExpandLastArg(args)) {
			const headArgs = printedArguments.slice(0, -1);
			if (headArgs.some(willBreak)) return allArgsBrokenOut();
			let lastArg;
			try {
				lastArg = path.call((arg) => print(arg, { expandLastArg: true }), "namedChildren", lastArgIndex);
			} catch (caught) {
				if (caught instanceof ArgExpansionBailout) return allArgsBrokenOut();
				throw caught;
			}
			if (willBreak(lastArg)) return [breakParent$1, conditionalGroup([[
				"(",
				...headArgs,
				group$4(lastArg, { shouldBreak: true }),
				")"
			], allArgsBrokenOut()])];
			return conditionalGroup([
				[
					"(",
					...headArgs,
					lastArg,
					")"
				],
				[
					"(",
					...headArgs,
					group$4(lastArg, { shouldBreak: true }),
					")"
				],
				allArgsBrokenOut()
			]);
		}
		return group$4(indentInParentheses(printedArguments), { shouldBreak: printedArguments.some(willBreak) || anyArgEmptyLine });
	},
	array_creation_expression(path, print) {
		const parts = ["new "];
		path.each((child) => {
			if (child.node.type === "annotation" || child.node.type === "marker_annotation") parts.push(print(child), " ");
		}, "namedChildren");
		parts.push(path.call(print, "typeNode"), ...path.map(print, "dimensionsNodes"));
		if (hasChild(path, "valueNode")) parts.push(" ", path.call(print, "valueNode"));
		return parts;
	},
	dimensions_expr(path, print) {
		return path.map((child) => child.node.type === "annotation" || child.node.type === "marker_annotation" ? [print(child), " "] : [
			"[",
			print(child),
			"]"
		], "namedChildren");
	},
	class_literal(path, print) {
		return [path.call(print, "namedChildren", 0), ".class"];
	},
	array_access: printMemberChain,
	method_reference(path, print) {
		return path.map(print, "children");
	},
	template_expression(path, print) {
		return [
			path.call(print, "template_processorNode"),
			".",
			path.call(print, "template_argumentNode")
		];
	},
	pattern(path, print) {
		return path.call(print, "namedChildren", 0);
	},
	type_pattern(path, print) {
		return join$4(" ", path.map(print, "children"));
	},
	record_pattern(path, print) {
		return path.map(print, "children");
	},
	record_pattern_body(path, print) {
		return group$4(indentInParentheses(join$4([",", line$4], path.map(print, "namedChildren"))));
	},
	record_pattern_component(path, print) {
		return join$4(" ", path.map(print, "children"));
	},
	guard(path, print) {
		const hasParentheses = path.node.namedChildren[0].type === "parenthesized_expression";
		const expression = path.call(print, "namedChildren", 0);
		return ["when ", hasParentheses ? expression : group$4([
			ifBreak("("),
			indent$4([softline$2, expression]),
			softline$2,
			ifBreak(")")
		])];
	}
};
function printLambdaExpressionSignature(path, options, print, args) {
	const parts = [];
	const parameters = path.call(print, "parametersNode");
	if (shouldPrintParamsWithoutParens(path, options)) parts.push(parameters);
	else {
		if (args != null && typeof args === "object" && ("expandLastArg" in args && args.expandLastArg === true || "expandFirstArg" in args && args.expandFirstArg === true)) {
			if (willBreak(parameters)) throw new ArgExpansionBailout();
			parts.push(group$4(removeLines(parameters)));
		} else parts.push(parameters);
		if (path.node.parametersNode.type === "identifier") {
			parts.unshift("(");
			parts.push(")");
		}
	}
	const dangling = printDanglingComments(path);
	if (dangling.length) parts.push(" ", dangling);
	return parts;
}
function mayBreakAfterShortPrefix(functionBody) {
	return functionBody.type === "array_creation_expression" || functionBody.type === "lambda_expression" || functionBody.type === "block";
}
function printArrowFunctionSignatures(path, { signatureDocs }) {
	if (signatureDocs.length === 1) return signatureDocs[0];
	const { node, parent } = path;
	if (node.fieldName !== "object" && parent?.type === "method_invocation" || parent?.type === "binary_expression") return group$4([
		signatureDocs[0],
		" ->",
		indent$4([line$4, join$4([" ->", line$4], signatureDocs.slice(1))])
	]);
	if (node.fieldName === "object" && parent?.type === "method_invocation") return group$4(join$4([" ->", line$4], signatureDocs));
	return group$4(indent$4(join$4([" ->", line$4], signatureDocs)));
}
function shouldPrintParamsWithoutParens(path, options) {
	if (options.arrowParens === "always") return false;
	if (options.arrowParens === "avoid") {
		const { node } = path;
		return canPrintParamsWithoutParens(node);
	}
	return false;
}
function canPrintParamsWithoutParens(node) {
	return node.parametersNode.type === "identifier" && !node.comments?.some(({ leading, trailing }) => !leading && !trailing) && !node.parametersNode.comments;
}
function printMemberChain(path, print, options) {
	const isExpressionStatement = path.parent?.type === "expression_statement";
	const printedNodes = [];
	function shouldInsertEmptyLineAfter(node) {
		const { originalText } = options;
		const nextCharIndex = getNextNonSpaceNonCommentCharacterIndex(originalText, node.end.index);
		if ((nextCharIndex ? originalText.charAt(nextCharIndex) : "") === ")") return nextCharIndex !== false && isNextLineEmpty(originalText, nextCharIndex + 1);
		return isNextLineEmpty(originalText, node.end.index);
	}
	function rec(path) {
		const { node } = path;
		if (hasType(path, "method_invocation") && hasChild(path, "objectNode")) {
			const hasTrailingEmptyLine = shouldInsertEmptyLineAfter(node);
			printedNodes.unshift({
				node,
				hasTrailingEmptyLine,
				printed: [printComments(path, printMethodInvocation(path, print)), hasTrailingEmptyLine ? hardline$3 : ""]
			});
			path.call(rec, "objectNode");
		} else if (hasType(path, "array_access")) {
			printedNodes.unshift({
				node,
				printed: printComments(path, printArrayAccess(path, print))
			});
			path.call(rec, "arrayNode");
		} else if (hasType(path, "field_access")) {
			printedNodes.unshift({
				node,
				printed: printComments(path, printFieldAccess(path, print))
			});
			path.call(rec, "objectNode");
		} else printedNodes.unshift({
			node,
			printed: print(path)
		});
	}
	const { node } = path;
	if (hasType(path, "method_invocation")) {
		printedNodes.unshift({
			node,
			printed: printComments(path, printMethodInvocation(path, print))
		});
		if (hasChild(path, "objectNode")) path.call(rec, "objectNode");
	} else if (hasType(path, "array_access")) {
		printedNodes.unshift({
			node,
			printed: printComments(path, printArrayAccess(path, print))
		});
		if (hasChild(path, "arrayNode")) path.call(rec, "arrayNode");
	} else if (hasType(path, "field_access")) {
		printedNodes.unshift({
			node,
			printed: printComments(path, printFieldAccess(path, print))
		});
		if (hasChild(path, "objectNode")) path.call(rec, "objectNode");
	}
	const danglingComments = printDanglingComments(path);
	if (danglingComments.length) printedNodes[0].printed = [
		...danglingComments,
		hardline$3,
		printedNodes[0].printed
	];
	const groups = [];
	let currentGroup = [printedNodes[0]];
	let i = 1;
	for (; i < printedNodes.length; ++i) if (printedNodes[i].node.type === "array_access") currentGroup.push(printedNodes[i]);
	else break;
	if (printedNodes[0].node.type !== "method_invocation") for (; i + 1 < printedNodes.length; ++i) if (printedNodes[i].node.type !== "method_invocation") currentGroup.push(printedNodes[i]);
	else break;
	groups.push(currentGroup);
	currentGroup = [];
	let hasSeenMethodInvocation = false;
	for (; i < printedNodes.length; ++i) {
		if (hasSeenMethodInvocation) {
			if (printedNodes[i].node.type === "array_access") {
				currentGroup.push(printedNodes[i]);
				continue;
			}
			groups.push(currentGroup);
			currentGroup = [];
			hasSeenMethodInvocation = false;
		}
		if (printedNodes[i].node.type === "method_invocation") hasSeenMethodInvocation = true;
		currentGroup.push(printedNodes[i]);
		if (printedNodes[i].node.comments?.some(({ trailing }) => trailing)) {
			groups.push(currentGroup);
			currentGroup = [];
			hasSeenMethodInvocation = false;
		}
	}
	if (currentGroup.length > 0) groups.push(currentGroup);
	function isFactory(name) {
		return /^[A-Z]|^[$_]+$/.test(name);
	}
	function isShort(name) {
		return name.length <= options.tabWidth;
	}
	function shouldNotWrap(groups) {
		const hasArrayAccess = groups[1][0]?.node.type === "array_access";
		if (groups[0].length === 1) {
			const firstNode = groups[0][0].node;
			return firstNode.type === "this" || firstNode.type === "identifier" && (isFactory(firstNode.value) || isExpressionStatement && isShort(firstNode.value) || hasArrayAccess);
		}
		const lastNode = groups[0].at(-1).node;
		return lastNode.type === "field_access" && lastNode.fieldNode.type === "identifier" && (isFactory(lastNode.fieldNode.value) || hasArrayAccess);
	}
	const shouldMerge = groups.length >= 2 && !groups[1][0].node.comments?.length && shouldNotWrap(groups);
	function printGroup(printedGroup) {
		return printedGroup.map((tuple) => tuple.printed);
	}
	function printIndentedGroup(groups) {
		if (groups.length === 0) return "";
		return indent$4([hardline$3, join$4(hardline$3, groups.map(printGroup))]);
	}
	const printedGroups = groups.map(printGroup);
	const oneLine = printedGroups;
	const cutoff = shouldMerge ? 3 : 2;
	const flatGroups = groups.flat();
	const nodeHasComment = flatGroups.some((node) => node.node.comments?.some(({ leading }) => leading)) || flatGroups.slice(0, -1).some((node) => node.node.comments?.some(({ trailing }) => trailing));
	if (groups.length <= cutoff && !nodeHasComment && !groups.some((g) => g.at(-1).hasTrailingEmptyLine)) return group$4(oneLine);
	const lastNodeBeforeIndent = groups[shouldMerge ? 1 : 0].at(-1).node;
	const shouldHaveEmptyLineBeforeIndent = lastNodeBeforeIndent.type !== "method_invocation" && shouldInsertEmptyLineAfter(lastNodeBeforeIndent);
	const expanded = [
		printGroup(groups[0]),
		shouldMerge ? groups.slice(1, 2).map(printGroup) : "",
		shouldHaveEmptyLineBeforeIndent ? hardline$3 : "",
		printIndentedGroup(groups.slice(shouldMerge ? 2 : 1))
	];
	const methodInvocations = printedNodes.map(({ node }) => node).filter((node) => node.type === "method_invocation");
	function lastGroupWillBreakAndOtherCallsHaveFunctionArguments() {
		const lastGroupNode = groups.at(-1).at(-1).node;
		const lastGroupDoc = printedGroups.at(-1);
		return lastGroupNode.type === "method_invocation" && willBreak(lastGroupDoc) && methodInvocations.slice(0, -1).some((node) => node.argumentsNode.namedChildren.some(({ type }) => type === "lambda_expression"));
	}
	let result;
	if (nodeHasComment || methodInvocations.length > 2 && methodInvocations.some((inv) => !inv.argumentsNode.namedChildren.every((arg) => isSimpleCallArgument(arg))) || printedGroups.slice(0, -1).some(willBreak) || lastGroupWillBreakAndOtherCallsHaveFunctionArguments()) result = group$4(expanded);
	else result = [willBreak(oneLine) || shouldHaveEmptyLineBeforeIndent ? breakParent$1 : "", conditionalGroup([oneLine, expanded])];
	return result;
}
function printMethodInvocation(path, print) {
	const parts = [];
	if (hasChild(path, "objectNode")) parts.push(".");
	if (path.node.children.filter(({ type }) => type === ".").length === 2) parts.push("super", ".");
	if (hasChild(path, "type_argumentsNode")) parts.push(path.call(print, "type_argumentsNode"));
	parts.push(path.call(print, "nameNode"), path.call(print, "argumentsNode"));
	return parts;
}
function printArrayAccess(path, print) {
	const index = path.call(print, "indexNode");
	return path.node.indexNode.type === "decimal_integer_literal" ? [
		"[",
		index,
		"]"
	] : group$4([
		"[",
		indent$4([softline$2, index]),
		softline$2,
		"]"
	]);
}
function printFieldAccess(path, print) {
	const parts = ["."];
	if (path.node.children.filter(({ type }) => type === ".").length === 2) parts.push("super.");
	parts.push(path.call(print, "fieldNode"));
	return parts;
}
/**
* For binary expressions to be consistent, we need to group
* subsequent operators with the same precedence level under a single
* group. Otherwise they will be nested such that some of them break
* onto new lines but not all. Operators with the same precedence
* level should either all break or not. Because we group them by
* precedence level and the AST is structured based on precedence
* level, things are naturally broken up correctly, i.e. `&&` is
* broken before `+`.
*/
function printBinaryExpressions(path, print, options, isInsideParentheses) {
	if (!hasType(path, "binary_expression")) return [group$4(print(path))];
	const { node } = path;
	let parts = [];
	if (node.leftNode.type === "binary_expression" && shouldFlatten(node.operatorNode.type, node.leftNode.operatorNode.type)) parts = path.call(() => printBinaryExpressions(path, print, options, isInsideParentheses), "leftNode");
	else parts.push(group$4(path.call(print, "leftNode")));
	const operator = node.operatorNode.type;
	const operatorDoc = path.call(print, "operatorNode");
	const rightContent = path.call(print, "rightNode");
	let right = options.experimentalOperatorPosition === "start" ? [
		line$4,
		operatorDoc,
		" ",
		rightContent
	] : [
		" ",
		operatorDoc,
		line$4,
		rightContent
	];
	const { parent } = path;
	const shouldBreak = node.leftNode.comments?.some(({ trailing, type }) => trailing && type === "line_comment") ?? false;
	if (shouldBreak || !(isInsideParentheses && logicalOperators.has(operator)) && (parent?.type !== node.type || logicalOperators.has(parent.operatorNode.value) !== logicalOperators.has(operator)) && node.leftNode.type !== node.type && node.rightNode.type !== node.type) right = group$4(right, { shouldBreak });
	parts.push(right);
	return parent?.type === "binary_expression" && needsParentheses(parent.operatorNode.type, operator) ? [
		"(",
		...parts,
		")"
	] : parts;
}
const logicalOperators = new Set(["||", "&&"]);
const equalityOperators = new Set(["==", "!="]);
const multiplicativeOperators = new Set([
	"*",
	"/",
	"%"
]);
const bitshiftOperators = new Set([
	">>",
	">>>",
	"<<"
]);
function shouldFlatten(parentOp, nodeOp) {
	if (getPrecedence(nodeOp) !== getPrecedence(parentOp)) return false;
	if (equalityOperators.has(parentOp) && equalityOperators.has(nodeOp)) return false;
	if (nodeOp === "%" && multiplicativeOperators.has(parentOp) || parentOp === "%" && multiplicativeOperators.has(nodeOp)) return false;
	if (nodeOp !== parentOp && multiplicativeOperators.has(nodeOp) && multiplicativeOperators.has(parentOp)) return false;
	if (bitshiftOperators.has(parentOp) && bitshiftOperators.has(nodeOp)) return false;
	return true;
}
const PRECEDENCE = new Map([
	["||"],
	["&&"],
	["|"],
	["^"],
	["&"],
	["==", "!="],
	[
		"<",
		">",
		"<=",
		">=",
		"instanceof"
	],
	[
		"<<",
		">>",
		">>>"
	],
	["+", "-"],
	[
		"*",
		"/",
		"%"
	]
].flatMap((operators, index) => operators.map((operator) => [operator, index])));
function getPrecedence(operator) {
	return PRECEDENCE.get(operator) ?? -1;
}
function needsParentheses(parentOperator, operator) {
	return operator === "&&" && parentOperator === "||" || [
		"|",
		"^",
		"&",
		"<<",
		">>",
		">>>"
	].includes(parentOperator) && getPrecedence(operator) > getPrecedence(parentOperator) || [operator, parentOperator].every((o) => ["==", "!="].includes(o)) || [operator, parentOperator].every((o) => [
		"<<",
		">>",
		">>>"
	].includes(o)) || operator === "*" && parentOperator === "/" || operator === "/" && parentOperator === "*" || operator === "%" && [
		"+",
		"-",
		"*",
		"/"
	].includes(parentOperator) || ["*", "/"].includes(operator) && parentOperator === "%";
}
function isSimpleCallArgument(node, depth = 2) {
	if (depth <= 0) return false;
	const isChildSimple = (child) => isSimpleCallArgument(child, depth - 1);
	if (node.type.endsWith("_literal") || node.type === "true" || node.type === "false" || node.type === "identifier" || node.type === "this") return true;
	if (node.type === "object_creation_expression" || node.type === "method_invocation") {
		if (node.type === "object_creation_expression" || !node.objectNode || isSimpleCallArgument(node.objectNode, depth)) {
			const args = node.argumentsNode.namedChildren;
			return args.length <= depth && args.every(isChildSimple);
		}
		return false;
	}
	if (node.type === "array_access") return isSimpleCallArgument(node.arrayNode, depth) && isSimpleCallArgument(node.indexNode, depth);
	if (node.type === "field_access") return isSimpleCallArgument(node.objectNode, depth) && isSimpleCallArgument(node.fieldNode, depth);
	if (node.type === "method_reference" || node.type === "unary_expression" || node.type === "update_expression") return isSimpleCallArgument(node.namedChildren[0], depth);
	return false;
}
function couldExpandArg(arg, lambdaChainRecursion = false) {
	if (arg.type === "array_creation_expression" && arg.valueNode && (arg.valueNode.namedChildren.length > 0 || arg.valueNode.comments)) return true;
	if (arg.type === "lambda_expression") {
		const { bodyNode: body } = arg;
		if (body.type === "block" || body.type === "array_creation_expression") return true;
		if (body.type === "lambda_expression" && couldExpandArg(body, true)) return true;
		if (!lambdaChainRecursion) {
			if (body.type === "ternary_expression") return true;
			if (body.type === "method_invocation" || body.type === "object_creation_expression") return true;
		}
	}
	return false;
}
function shouldExpandLastArg(args) {
	const lastArg = args.at(-1);
	const penultimateArg = args.at(-2);
	return !lastArg.comments?.some(({ leading }) => leading) && !lastArg.comments?.some(({ trailing }) => trailing) && couldExpandArg(lastArg) && (!penultimateArg || penultimateArg.type !== lastArg.type) && (args.length !== 2 || penultimateArg.type !== "lambda_expression");
}
function shouldExpandFirstArg(args) {
	if (args.length !== 2) return false;
	const [firstArg, secondArg] = args;
	return !firstArg.comments && firstArg.type === "lambda_expression" && firstArg.bodyNode.type === "block" && secondArg.type !== "lambda_expression" && secondArg.type !== "ternary_expression" && isHopefullyShortCallArgument(secondArg) && !couldExpandArg(secondArg);
}
function isHopefullyShortCallArgument(node) {
	if (node.type === "parenthesized_expression") return isHopefullyShortCallArgument(node.namedChildren[0]);
	if ((node.type === "method_invocation" || node.type === "object_creation_expression") && node.argumentsNode.namedChildren.length > 1) return false;
	if (node.type === "binary_expression") return isSimpleCallArgument(node.leftNode, 1) && isSimpleCallArgument(node.rightNode, 1);
	return isSimpleCallArgument(node);
}
var ArgExpansionBailout = class extends Error {
	constructor(..._args) {
		super(..._args);
		this.name = "ArgExpansionBailout";
	}
};
//#endregion
//#region src/printers/interfaces.ts
const { group: group$3, indent: indent$3, join: join$3, line: line$3 } = prettier_doc.builders;
var interfaces_default = {
	interface_declaration(path, print) {
		const parts = ["interface ", path.call(print, "nameNode")];
		const extendsInterfacesIndex = path.node.namedChildren.findIndex(({ type }) => type === "extends_interfaces");
		const hasExtendsInterfaces = extendsInterfacesIndex !== -1;
		const hasPermits = hasChild(path, "permitsNode");
		const hasMultipleClauses = hasExtendsInterfaces && hasPermits;
		const hasTypeParameters = hasChild(path, "type_parametersNode");
		if (hasTypeParameters) {
			const typeParameters = path.call(print, "type_parametersNode");
			parts.push(hasMultipleClauses ? group$3(indent$3(typeParameters)) : typeParameters);
		}
		if (hasExtendsInterfaces || hasPermits) {
			const separator = hasTypeParameters && !hasMultipleClauses ? " " : line$3;
			const clauses = [];
			if (hasExtendsInterfaces) clauses.push(separator, path.call(print, "namedChildren", extendsInterfacesIndex));
			if (hasPermits) clauses.push(separator, path.call(print, "permitsNode"));
			const hasBody = path.node.bodyNode.namedChildren.length > 0;
			const clauseGroup = [hasTypeParameters && !hasMultipleClauses ? clauses : indent$3(clauses), hasBody ? separator : " "];
			parts.push(hasMultipleClauses ? clauseGroup : group$3(clauseGroup));
		} else parts.push(" ");
		return [
			...printModifiers(path, print, "declarationOnly"),
			group$3(parts),
			path.call(print, "bodyNode")
		];
	},
	extends_interfaces(path, print) {
		const typeListIndex = path.node.namedChildren.findIndex(({ type }) => type === "type_list");
		return group$3(["extends", indent$3([line$3, path.call(print, "namedChildren", typeListIndex)])]);
	},
	interface_body(path, print) {
		return printBlock(path, printBodyDeclarations(path, print));
	},
	constant_declaration: printVariableDeclaration,
	annotation_type_declaration(path, print) {
		const parts = printModifiers(path, print);
		parts.push("@interface ", path.call(print, "nameNode"), " ", path.call(print, "bodyNode"));
		return parts;
	},
	annotation_type_body(path, print) {
		return printBlock(path, printBodyDeclarations(path, print));
	},
	annotation_type_element_declaration(path, print) {
		const parts = printModifiers(path, print);
		parts.push(path.call(print, "typeNode"), " ", path.call(print, "nameNode"), "()");
		if (hasChild(path, "dimensionsNode")) parts.push(path.call(print, "dimensionsNode"));
		if (hasChild(path, "valueNode")) parts.push(" default ", path.call(print, "valueNode"));
		parts.push(";");
		return parts;
	},
	annotation(path, print) {
		return [
			"@",
			path.call(print, "nameNode"),
			path.call(print, "argumentsNode")
		];
	},
	marker_annotation(path, print) {
		return ["@", path.call(print, "nameNode")];
	},
	annotation_argument_list(path, print) {
		const args = path.map(print, "namedChildren");
		return args.length === 1 && path.node.namedChildren[0].type === "element_value_array_initializer" ? [
			"(",
			args[0],
			")"
		] : group$3(indentInParentheses(join$3([",", line$3], args)));
	},
	element_value_pair(path, print) {
		return group$3([
			path.call(print, "keyNode"),
			" = ",
			path.call(print, "valueNode")
		]);
	},
	element_value_array_initializer: printArrayInitializer
};
//#endregion
//#region src/printers/lexical-structure.ts
const { group: group$2, hardline: hardline$2, indent: indent$2, join: join$2, softline: softline$1 } = prettier_doc.builders;
var lexical_structure_default = {
	string_literal(path, print) {
		if (path.node.namedChildren.some(({ type }) => type === "string_interpolation") || path.node.children[0].value === "\"") return path.map(print, "children");
		return printTextBlock(path, join$2(hardline$2, textBlockContents(path.node).split("\n")));
	},
	string_fragment: printValue,
	multiline_string_fragment: printValue,
	string_interpolation(path, print) {
		const expressionType = path.node.namedChildren[0].type;
		const expression = path.call(print, "namedChildren", 0);
		return expressionType === "binary_expression" || expressionType === "ternary_expression" ? group$2([
			"\\{",
			indent$2([softline$1, expression]),
			softline$1,
			"}"
		]) : [
			"\\{",
			expression,
			"}"
		];
	},
	escape_sequence: printValue,
	character_literal: printValue,
	binary_integer_literal: printValue,
	decimal_integer_literal: printValue,
	hex_integer_literal: printValue,
	octal_integer_literal: printValue,
	decimal_floating_point_literal: printValue,
	hex_floating_point_literal: printValue,
	null_literal: printValue,
	true: printValue,
	false: printValue,
	this: printValue,
	super: printValue,
	underscore_pattern: printValue,
	asterisk: printValue
};
//#endregion
//#region src/printers/names.ts
var names_default = {
	identifier: printValue,
	type_identifier: printValue,
	scoped_identifier(path, print) {
		return [
			path.call(print, "scopeNode"),
			".",
			path.call(print, "nameNode")
		];
	},
	scoped_type_identifier(path, print) {
		return path.map((child) => child.node.type === "annotation" || child.node.type === "marker_annotation" ? [print(child), " "] : print(child), "children");
	}
};
//#endregion
//#region src/printers/packages-and-modules.ts
const { group: group$1, hardline: hardline$1, indent: indent$1, join: join$1, line: line$2 } = prettier_doc.builders;
var packages_and_modules_default = {
	program(path, print) {
		if (!path.node.namedChildren.length) return [...printDanglingComments(path), hardline$1];
		const parts = [];
		if (path.node.namedChildren[0].type !== "import_declaration") parts.push(path.call(print, "namedChildren", 0));
		const staticImports = [];
		const imports = [];
		const otherDeclarations = [];
		path.each((child) => {
			const doc = print(child);
			if (child.node.type === "import_declaration") {
				const names = extractNames(child.node.namedChildren[0]);
				if (child.node.namedChildren.some(({ type }) => type === "asterisk")) names.push("*");
				(child.node.children[1].type === "static" ? staticImports : imports).push({
					doc,
					names
				});
			} else if (!child.isFirst) otherDeclarations.push(doc);
		}, "namedChildren");
		parts.push(...[staticImports, imports].filter((imports) => imports.length).map((imports) => join$1(hardline$1, imports.sort(compareFqn).map(({ doc }) => doc))), ...otherDeclarations);
		return [...join$1([hardline$1, hardline$1], parts), hardline$1];
	},
	package_declaration(path, print) {
		const annotations = [];
		const identifier = [];
		path.each((child) => {
			switch (child.node.type) {
				case "annotation":
				case "marker_annotation":
					annotations.push(print(child));
					break;
				case "identifier":
				case "scoped_identifier":
					identifier.push(print(child));
					break;
			}
		}, "namedChildren");
		return join$1(hardline$1, [...annotations, [
			"package ",
			...identifier,
			";"
		]]);
	},
	import_declaration(path, print) {
		const declaration = ["import "];
		if (path.node.children.some(({ type }) => type === "static")) declaration.push("static ");
		const identifierIndex = path.node.namedChildren.findIndex(({ type }) => type === "identifier" || type === "scoped_identifier");
		declaration.push(path.call(print, "namedChildren", identifierIndex));
		if (path.node.namedChildren.some(({ type }) => type === "asterisk")) declaration.push(".*");
		declaration.push(";");
		return declaration;
	},
	module_declaration(path, print) {
		const parts = [];
		path.each((child) => {
			if (child.node.type === "annotation" || child.node.type === "marker_annotation") parts.push(print(child));
		}, "namedChildren");
		if (path.node.children.some(({ type }) => type === "open")) parts.push("open");
		parts.push("module", path.call(print, "nameNode"), path.call(print, "bodyNode"));
		return join$1(" ", parts);
	},
	module_body(path, print) {
		return printBlock(path, path.map((child) => child.previous && child.node.start.row > child.previous.end.row + 1 ? [hardline$1, print(child)] : print(child), "namedChildren"));
	},
	requires_module_directive(path, print) {
		const parts = ["requires"];
		path.each((child) => {
			if (child.node.type === "requires_modifier") parts.push(print(child));
		}, "namedChildren");
		parts.push(path.call(print, "moduleNode"));
		return [...join$1(" ", parts), ";"];
	},
	exports_module_directive: printToModuleNamesDirective,
	opens_module_directive: printToModuleNamesDirective,
	uses_module_directive(path, print) {
		return [
			"uses ",
			path.call(print, "typeNode"),
			";"
		];
	},
	provides_module_directive(path, print) {
		const [provided, ...providers] = path.map(print, "namedChildren");
		return [
			"provides ",
			provided,
			group$1(indent$1([line$2, group$1(indent$1([
				"with",
				line$2,
				...join$1([",", line$2], providers)
			]))])),
			";"
		];
	},
	requires_modifier: printValue
};
function extractNames(node) {
	return node.type === "identifier" ? [node.value] : [...extractNames(node.scopeNode), node.nameNode.value];
}
function compareFqn({ names: a }, { names: b }) {
	const minParts = Math.min(a.length, b.length);
	for (let i = 0; i < minParts; i++) {
		const imageA = a[i];
		const imageB = b[i];
		if (imageA < imageB) return -1;
		else if (imageA > imageB) return 1;
	}
	return a.length - b.length;
}
function printToModuleNamesDirective(path, print) {
	const directive = [
		path.node.type === "exports_module_directive" ? "exports" : "opens",
		" ",
		path.call(print, "packageNode")
	];
	if (path.node.modulesNodes.length) {
		const moduleNames = join$1([",", line$2], path.map(print, "modulesNodes"));
		directive.push(group$1(indent$1([line$2, group$1(indent$1([
			"to",
			line$2,
			...moduleNames
		]))])));
	}
	directive.push(";");
	return directive;
}
//#endregion
//#region src/printers/types-values-and-variables.ts
const { group, indent, join, line: line$1, softline } = prettier_doc.builders;
var types_values_and_variables_default = {
	boolean_type: printValue,
	integral_type: printValue,
	floating_point_type: printValue,
	void_type: printValue,
	array_type(path, print) {
		return [path.call(print, "elementNode"), path.call(print, "dimensionsNode")];
	},
	annotated_type(path, print) {
		return join(" ", path.map(print, "children"));
	},
	dimensions(path, print) {
		return path.map((child) => {
			if (child.node.isNamed) return [
				...child.isFirst ? [" "] : [],
				print(child),
				" "
			];
			return child.node.value;
		}, "children");
	},
	type_parameter(path, print) {
		return join(" ", path.map(print, "children"));
	},
	type_bound(path, print) {
		const [firstType, ...restTypes] = path.map(print, "namedChildren");
		const bound = ["extends ", firstType];
		if (restTypes.length) bound.push(group(indent([
			line$1,
			"& ",
			...join([line$1, "& "], restTypes)
		])));
		return bound;
	},
	type_arguments(path, print) {
		const types = path.map(print, "namedChildren");
		return types.length ? group([
			"<",
			indent([softline, join([",", line$1], types)]),
			softline,
			">"
		]) : "<>";
	},
	wildcard(path, print) {
		return join(" ", path.map(print, "children"));
	}
};
//#endregion
//#region src/printers/index.ts
const printersByNodeType = {
	ERROR(path) {
		throw new Error(`Failed to parse: "${printValue(path)}"`);
	},
	...arrays_default,
	...blocks_and_statements_default,
	...classes_default,
	...expressions_default,
	...interfaces_default,
	...lexical_structure_default,
	...names_default,
	...packages_and_modules_default,
	...types_values_and_variables_default
};
function printerForNodeType(type) {
	return printersByNodeType[type];
}
//#endregion
//#region src/printer.ts
var printer_default = {
	print(path, options, print, args) {
		return hasJavaNode(path) ? printerForNodeType(path.node.type)(path, print, options, args) : printValue(path);
	},
	embed(path) {
		return hasType(path, "string_literal") ? embedTextBlock(path) : null;
	},
	hasPrettierIgnore(path) {
		return path.node.comments?.some(isPrettierIgnore) === true || canAttachComment(path.node) && isFullyBetweenPrettierIgnore(path);
	},
	canAttachComment,
	isBlockComment(node) {
		return node.type === "block_comment";
	},
	willPrintOwnComments,
	printComment(commentPath) {
		return printComment(commentPath.node);
	},
	getCommentChildNodes(node) {
		return node.isNamed ? node.children : [];
	},
	handleComments: {
		ownLine: handleLineComment,
		endOfLine: handleLineComment,
		remaining: handleRemainingComment
	},
	getVisitorKeys() {
		return ["namedChildren"];
	}
};
function hasJavaNode(path) {
	return path.node.isNamed;
}
//#endregion
//#region src/comments.ts
const { hasNewline, isPreviousLineEmpty, skipNewline, skipSpaces } = prettier.util;
const { breakParent, hardline, line, lineSuffix } = prettier_doc.builders;
const prettierIgnoreRangesByTree = /* @__PURE__ */ new WeakMap();
function determinePrettierIgnoreRanges(tree) {
	const { comments } = tree;
	if (!comments) return;
	const ranges = comments.filter(({ value }) => /^\/(?:\/\s*(?:prettier-ignore-(?:start|end)|@formatter:(?:off|on))\s*|\*\s*(?:prettier-ignore-(?:start|end)|@formatter:(?:off|on))\s*\*\/)$/.test(value)).reduce((ranges, { value, start }) => {
		const previous = ranges.at(-1);
		if (value.includes("start") || value.includes("off")) {
			if (previous?.end !== Infinity) ranges.push({
				start: start.index,
				end: Infinity
			});
		} else if (previous?.end === Infinity) previous.end = start.index;
		return ranges;
	}, new Array());
	prettierIgnoreRangesByTree.set(tree, ranges);
}
function isFullyBetweenPrettierIgnore(path) {
	const { node, root } = path;
	const start = parser_default.locStart(node);
	const end = parser_default.locEnd(node);
	return prettierIgnoreRangesByTree.get(root)?.some((range) => range.start < start && end < range.end) === true;
}
function isPrettierIgnore(comment) {
	return /^(\/\/\s*prettier-ignore|\/\*\s*prettier-ignore\s*\*\/)$/.test(comment.value);
}
function willPrintOwnComments(path) {
	return isMember(path.node) && !printer_default.hasPrettierIgnore(path);
}
function canAttachComment(node) {
	if (!node.isNamed) return isBinaryOperator(node);
	switch (node.type) {
		case "enum_body_declarations":
		case "escape_sequence":
		case "formal_parameters":
		case "modifier":
		case "multiline_string_fragment":
		case "parenthesized_expression":
		case "program":
		case "string_fragment":
		case "visibility": return false;
		default: return true;
	}
}
function handleLineComment(commentNode, _, options) {
	return [
		handleBinaryExpressionComments,
		handleFqnOrRefTypeComments,
		handleIfStatementComments,
		handleJumpStatementComments,
		handleLabeledStatementComments,
		handleMemberChainComments,
		handleModifiersComments,
		handleNameComments,
		handleTernaryExpressionComments,
		handleTryStatementComments
	].some((fn) => fn(commentNode, options));
}
function handleRemainingComment(commentNode) {
	return [
		handleFqnOrRefTypeComments,
		handleNameComments,
		handleJumpStatementComments
	].some((fn) => fn(commentNode));
}
function handleBinaryExpressionComments(commentNode, options) {
	const { enclosingNode, precedingNode, followingNode } = commentNode;
	if (enclosingNode?.type === "binary_expression") {
		if (isBinaryOperator(followingNode)) {
			if (options.experimentalOperatorPosition === "start") prettier.util.addLeadingComment(followingNode, commentNode);
			else prettier.util.addTrailingComment(followingNode, commentNode);
			return true;
		} else if (options.experimentalOperatorPosition === "start" && isBinaryOperator(precedingNode)) {
			prettier.util.addLeadingComment(precedingNode, commentNode);
			return true;
		}
	}
	return false;
}
function handleFqnOrRefTypeComments(commentNode) {
	const { enclosingNode, followingNode } = commentNode;
	if (enclosingNode?.type === "scoped_type_identifier" && followingNode) {
		prettier.util.addLeadingComment(followingNode, commentNode);
		return true;
	}
	return false;
}
function handleIfStatementComments(commentNode) {
	const { enclosingNode, precedingNode } = commentNode;
	if (enclosingNode?.type === "if_statement" && precedingNode?.fieldName === "consequence") {
		prettier.util.addDanglingComment(enclosingNode, commentNode, void 0);
		return true;
	}
	return false;
}
function handleJumpStatementComments(commentNode) {
	const { enclosingNode, precedingNode, followingNode } = commentNode;
	if (enclosingNode && !precedingNode && !followingNode && (enclosingNode.type === "break_statement" || enclosingNode.type === "continue_statement" || enclosingNode.type === "return_statement")) {
		prettier.util.addTrailingComment(enclosingNode, commentNode);
		return true;
	}
	return false;
}
function handleLabeledStatementComments(commentNode) {
	const { enclosingNode, precedingNode } = commentNode;
	if (enclosingNode?.type === "labeled_statement" && precedingNode?.type === "identifier") {
		prettier.util.addLeadingComment(precedingNode, commentNode);
		return true;
	}
	return false;
}
function handleMemberChainComments(commentNode) {
	const { enclosingNode, precedingNode, followingNode } = commentNode;
	if (precedingNode && (enclosingNode?.type === "field_access" || enclosingNode?.type === "method_invocation" && precedingNode.end.row < commentNode.start.row) && precedingNode === enclosingNode.objectNode) {
		prettier.util.addLeadingComment(enclosingNode, commentNode);
		return true;
	} else if (followingNode && isMember(followingNode) && (!precedingNode || precedingNode !== getMemberObject(followingNode) && precedingNode.end.row < commentNode.start.row) && !isPrettierIgnore(commentNode)) {
		prettier.util.addDanglingComment(followingNode, commentNode, void 0);
		return true;
	}
	return false;
}
function handleModifiersComments(commentNode) {
	const { precedingNode } = commentNode;
	if (precedingNode?.type === "annotation" || precedingNode?.type === "marker_annotation" || precedingNode?.type === "modifiers") {
		prettier.util.addTrailingComment(precedingNode, commentNode);
		return true;
	}
	return false;
}
function handleNameComments(commentNode) {
	const { enclosingNode, precedingNode } = commentNode;
	if (enclosingNode && precedingNode?.type === "identifier" && (enclosingNode.type === "scoped_identifier" || enclosingNode.type === "module_declaration" || enclosingNode.type === "package_declaration" || enclosingNode.type === "scoped_type_identifier")) {
		prettier.util.addTrailingComment(precedingNode, commentNode);
		return true;
	}
	return false;
}
function handleTernaryExpressionComments(commentNode) {
	const { enclosingNode, precedingNode, followingNode } = commentNode;
	if (enclosingNode?.type === "ternary_expression" && precedingNode?.isNamed && followingNode?.isNamed && precedingNode.end.row < commentNode.start.row && commentNode.end.row < followingNode.start.row) {
		prettier.util.addLeadingComment(followingNode, commentNode);
		return true;
	}
	return false;
}
function handleTryStatementComments(commentNode) {
	const { enclosingNode, followingNode } = commentNode;
	if (enclosingNode && (enclosingNode.type === "catch_clause" || enclosingNode.type === "try_statement" || enclosingNode.type === "try_with_resources_statement") && followingNode?.isNamed) {
		const block = followingNode.type === "catch_clause" ? followingNode.bodyNode : followingNode.type === "finally_clause" ? followingNode.namedChildren[0] : null;
		if (!block) return false;
		const blockStatement = block.namedChildren.at(0);
		if (blockStatement) prettier.util.addLeadingComment(blockStatement, commentNode);
		else prettier.util.addDanglingComment(block, commentNode, void 0);
		return true;
	}
	return false;
}
function isMember(node) {
	return node.type === "array_access" || node.type === "field_access" || node.type === "method_invocation";
}
function getMemberObject(node) {
	return node.type === "array_access" ? node.arrayNode : node.objectNode;
}
const binaryOperators = new Set([
	"<<",
	">>",
	">>>",
	"instanceof",
	"<=",
	">=",
	"==",
	"-",
	"+",
	"&&",
	"&",
	"^",
	"!=",
	"||",
	"|",
	"*",
	"/",
	"%"
]);
function isBinaryOperator(node) {
	return node !== void 0 && binaryOperators.has(node.type);
}
function printLeadingComment(path) {
	const comment = path.node;
	comment.printed = true;
	const parts = [printComment(comment)];
	const originalText = path.root.value;
	if (comment.type === "block_comment") {
		let lineBreak = " ";
		if (hasNewline(originalText, parser_default.locEnd(comment))) if (hasNewline(originalText, parser_default.locStart(comment), { backwards: true })) lineBreak = hardline;
		else lineBreak = line;
		parts.push(lineBreak);
	} else parts.push(hardline);
	const index = skipNewline(originalText, skipSpaces(originalText, parser_default.locEnd(comment)));
	if (index !== false && hasNewline(originalText, index)) parts.push(hardline);
	return parts;
}
function printTrailingComment(path, previousComment) {
	const comment = path.node;
	comment.printed = true;
	const printed = printComment(comment);
	const originalText = path.root.value;
	const isBlock = comment.type === "block_comment";
	if (previousComment?.hasLineSuffix && !previousComment?.isBlock || hasNewline(originalText, parser_default.locStart(comment), { backwards: true })) return {
		doc: lineSuffix([
			hardline,
			isPreviousLineEmpty(originalText, parser_default.locStart(comment)) ? hardline : "",
			printed
		]),
		isBlock,
		hasLineSuffix: true
	};
	if (!isBlock || previousComment?.hasLineSuffix) return {
		doc: [lineSuffix([" ", printed]), breakParent],
		isBlock,
		hasLineSuffix: true
	};
	return {
		doc: [" ", printed],
		isBlock,
		hasLineSuffix: false
	};
}
function printLeadingComments(path) {
	if (!hasChild(path, "comments")) return [];
	const docs = [];
	path.each((path) => {
		const { node: comment } = path;
		if (!comment.leading) return;
		docs.push(printLeadingComment(path));
	}, "comments");
	return docs;
}
function printTrailingComments(path) {
	if (!hasChild(path, "comments")) return [];
	const docs = [];
	let printedTrailingComment;
	path.each((path) => {
		const { node: comment } = path;
		if (!comment.trailing) return;
		printedTrailingComment = printTrailingComment(path, printedTrailingComment);
		docs.push(printedTrailingComment.doc);
	}, "comments");
	return docs;
}
function printCommentsSeparately(path) {
	return {
		leading: printLeadingComments(path),
		trailing: printTrailingComments(path)
	};
}
function printComments(path, doc) {
	const leading = printLeadingComments(path);
	const trailing = printTrailingComments(path);
	return leading.length || trailing.length ? [
		leading,
		doc,
		trailing
	] : doc;
}
//#endregion
//#region src/parser.ts
var parser_default = {
	async parse(text) {
		const tree = (await parser).parse(text);
		const { rootNode } = tree;
		if (rootNode.hasError) throw new Error("Failed to parse: " + rootNode);
		const ast = processTree(rootNode);
		determinePrettierIgnoreRanges(ast);
		tree.delete();
		return ast;
	},
	astFormat: "java",
	hasPragma(text) {
		return /^\/\*\*\n\s+\*\s@(format|prettier)\n\s+\*\//.test(text);
	},
	locStart(node) {
		return node.start.index;
	},
	locEnd(node) {
		return node.end.index;
	}
};
const parser = (async () => {
	await web_tree_sitter.Parser.init();
	const parser = new web_tree_sitter.Parser();
	const Java = await web_tree_sitter.Language.load(new URL("./tree-sitter-java_orchard.wasm", require("url").pathToFileURL(__filename).href));
	parser.setLanguage(Java);
	return parser;
})();
function processTree(node, fieldName = null, comments) {
	const { type, isNamed, text: value, startPosition, endPosition } = node;
	const javaNode = {
		type,
		isNamed,
		value,
		start: {
			index: node.startIndex,
			row: startPosition.row,
			column: startPosition.column
		},
		end: {
			index: node.endIndex,
			row: endPosition.row,
			column: endPosition.column
		},
		children: [],
		namedChildren: [],
		fieldName
	};
	if (!comments) comments = javaNode.comments = [];
	if (!javaNode.isNamed) return javaNode;
	const multiFields = multiFieldsByType[node.type];
	if (multiFields) Object.keys(multiFields).forEach((name) => javaNode[`${name}Nodes`] = []);
	node.children.forEach((child, index) => {
		const { type, text: value, startPosition, endPosition } = child;
		if (type === "block_comment" || type === "line_comment") comments.push({
			type,
			value,
			start: {
				index: child.startIndex,
				row: startPosition.row,
				column: startPosition.column
			},
			end: {
				index: child.endIndex,
				row: endPosition.row,
				column: endPosition.column
			},
			leading: false,
			trailing: false,
			printed: false
		});
		else {
			const fieldName = node.fieldNameForChild(index);
			const javaChild = processTree(child, fieldName, comments);
			javaNode.children.push(javaChild);
			if (javaChild.isNamed) javaNode.namedChildren.push(javaChild);
			if (fieldName) if (multiFields?.[fieldName]) javaNode[`${fieldName}Nodes`].push(javaChild);
			else javaNode[`${fieldName}Node`] = javaChild;
		}
	});
	return javaNode;
}
//#endregion
//#region src/index.ts
var src_default = {
	languages: [{
		name: "Java",
		parsers: ["java"],
		group: "Java",
		tmScope: "source.java",
		aceMode: "java",
		codemirrorMode: "clike",
		codemirrorMimeType: "text/x-java",
		extensions: [".java"],
		linguistLanguageId: 181,
		vscodeLanguageIds: ["java"]
	}],
	parsers: { java: parser_default },
	printers: { java: printer_default },
	options: options_default,
	defaultOptions: { arrowParens: "avoid" }
};
//#endregion
module.exports = src_default;
